lib/radiator_web/controllers/api/collaborator_controller.ex

```diff
@@ -0,0 +1,80 @@
+defmodule RadiatorWeb.Api.CollaboratorController do
+  use RadiatorWeb, :controller
+  use Radiator.Constants
+
+  alias Radiator.Directory.Editor
+
+  alias Radiator.Directory.Collaborator
+
+  action_fallback RadiatorWeb.Api.FallbackController
+
+  def create(conn, %{"username" => username, "permission" => permission} = params) do
```

Comment:
Feels too deeply nested to me. Here is a version that uses only one `with` block (for illustration, I didn't run it). However for finer grained control you would need the tagging as discussed before and that might become unreadable as well. The admin version of the controller is an even more interesting example as there are more failure handling cases already. Your call.

```elixir
  def create(conn, %{"username" => username, "permission" => permission} = params) do
    with actor <- current_user(conn),
         {:ok, subject} <- current_subject(actor, params),
         user when not is_nil(user) <- Radiator.Auth.Register.get_user_by_name(username),
         collaborator <- build_collaborator(user, permission, subject),
         {:ok, collaborator} <- Editor.add_collaborator(actor, collaborator) do
      conn
      |> put_status(:created)
      |> render("show.json", %{collaborator: collaborator})
    else
      _ -> {:error, :not_found}
    end
  end

  defp build_collaborator(user, permission, subject) do
    %Collaborator{
      user: user,
      permission: String.to_existing_atom(permission),
      subject: subject
    }
  end
```

Reply:
Well, you are killing the not authorized errors though. Get the point, Took the idea and reworked it a bit.